streams/src/main/java/org/apache/kafka/streams/processor/internals/ChangelogReader.java

```diff
@@ -37,7 +37,7 @@
      * Restore all registered state stores by reading from their changelogs.
      * @return all topic partitions that have been restored
      */
-    Collection<TopicPartition> restore();
+    Collection<TopicPartition> restore(final AssignedTasks active);
```

Comment:
Instead of passing the whole AssignedTasks object and hence introduce this penetrated dependency, could we instead throw the MigratedException as is, and only add the task information in `updateNewAndRestoringTasks` and re-throw? Then the exception's message function would depend on whether or not the task field is null.

Reply:
By doing so we do not need to change the current APIs at all.

Reply:
I understand your indention and I don't like the API change either. However, I am not sure if this would simplify the code. We would need to get the `TopicPartition` out of the exception, and alter the exception to add more information. Overall, it seems be rather confusing to understand what going on. We should have the information we need when raising the exception -- this keep code "local" and not spread out over multiple classes.

Reply:
I think you should create an interface rather than passing in the `AssignedTasks` class, it seems you only need a single method. The contract should be on an interface rather than the class.